fs/copy_linux.go

```diff
@@ -49,22 +49,31 @@ func copyFileContent(dst, src *os.File) error {
 		return errors.Wrap(err, "unable to stat source")
 	}
 
-	n, err := unix.CopyFileRange(int(src.Fd()), nil, int(dst.Fd()), nil, int(st.Size()), 0)
-	if err != nil {
-		if err != unix.ENOSYS && err != unix.EXDEV {
-			return errors.Wrap(err, "copy file range failed")
-		}
+	var written int64
+	size := st.Size()
 
-		buf := bufferPool.Get().(*[]byte)
-		_, err = io.CopyBuffer(dst, src, *buf)
-		bufferPool.Put(buf)
-		return err
-	}
+	first := true
+	for written < size {
+		n, err := unix.CopyFileRange(int(src.Fd()), nil, int(dst.Fd()), nil, int(size-written), 0)
+		if err != nil {
+			if (err != unix.ENOSYS && err != unix.EXDEV) || !first {
+				return errors.Wrap(err, "copy file range failed")
+			}
 
-	if int64(n) != st.Size() {
-		return errors.Wrapf(err, "short copy: %d of %d", int64(n), st.Size())
+			buf := bufferPool.Get().(*[]byte)
+			_, err = io.CopyBuffer(dst, src, *buf)
+			bufferPool.Put(buf)
+			return errors.Wrap(err, "userspace copy failed")
+		}
+		if n == 0 {
+			break
+		}
+		first = false
+		written += int64(n)
+	}
+	if written != size {
```

Comment:
this surely cant happen?

Reply:
i.e., I guess only if there was a bug in the kernel in the go syscall implementation.

Reply:
its not really worth checking for random kernel bugs